rfq/convert.go

```diff
@@ -0,0 +1,89 @@
+package rfq
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"golang.org/x/exp/constraints"
+)
+
+// FixedPoint is used to represent fixed point arithmetic for currency related
+// calculations. A fixed point consists of a value, and a scale. The value is
+// the integer representation of the number. The scale is used to represent the
+// fractional/decimal component.
+type FixedPoint[T constraints.Unsigned] struct {
+	// Value is the value of the FixedPoint integer.
+	Value T
+
+	// Scale is used to represent the fractional component. This always
+	// represents a power of 10. Eg: a scale value of 2 (two decimal places)
+	// maps to a multiplication by 100.
+	Scale int
+}
+
+// String returns the string version of the fixed point value.
+func (f FixedPoint[T]) String() string {
+	value := float64(f.Value) / math.Pow10(f.Scale)
+	return fmt.Sprintf("%.*f", f.Scale, value)
+}
+
+// ScaleTo returns a new FixedPoint that is scaled up or down to the given
+// scale.
+func (f FixedPoint[T]) ScaleTo(newScale int) FixedPoint[T] {
+	scaleDiff := newScale - f.Scale
+	multiplier := math.Pow10(scaleDiff)
+	newValue := float64(f.Value) * multiplier
+
+	return FixedPoint[T]{
+		Value: T(newValue),
+		Scale: newScale,
+	}
+}
+
+// MilliSatoshiToUnits converts the given milli-satoshi amount to units using
+// the given price in units per bitcoin as a fixed point in the asset's desired
+// resolution (scale equal to decimal display).
+func MilliSatoshiToUnits(milliSat lnwire.MilliSatoshi,
+	unitsPerBtc FixedPoint[uint64]) uint64 {
+
+	oneBtcInMilliSatF := toBigFloat(btcutil.SatoshiPerBitcoin * 1_000)
+	priceUnitsPerBtcF := toBigFloat(unitsPerBtc.Value)
+
+	milliSatPerUnitF := new(big.Float)
+	milliSatPerUnitF.Quo(oneBtcInMilliSatF, priceUnitsPerBtcF)
```

Comment:
I went to convert things to just use integers, this operation may result in a value smaller than one, when the the value of 1 unit <<< 1 msat. 

Will try to generalize the way the computation in the gist (`mSatToUSD`) works to emulate this behvaior.

Reply:
> I went to convert things to just use integers, this operation may result in a value smaller than one, when the the value of 1 unit <<< 1 msat.

The way I think about it, is we do `unitsPerMsat`, then that value is always a whole number.

Reply:
Actually scratch the above, I think I worked it out re the scale values, etc.